test.lua

```diff
@@ -781,6 +781,68 @@ do
     end
 end
 
+---- REPL tests ----
+print("Running tests for REPL completion...")
+local wrapRepl = function()
+    local replComplete
+    local replSend = coroutine.wrap(function()
+        local output = {}
+        fennel.repl({
+            readChunk = function()
+                local chunk = coroutine.yield(output)
+                output = {}
+                return chunk and chunk .. '\n' or nil
+            end,
+            onValues = function(xs)
+                table.insert(output, xs)
+            end,
+            registerCompleter = function(completer)
+                replComplete = function(input)
+                    local res = completer(input)
+                    table.sort(res)
+                    return res
+                end
+            end,
+            pp = function(x) return x end,
+        })
+    end) replSend()
+    return replSend, replComplete
+end
+
+-- Skip REPL tests in non-JIT Lua 5.1 only to avoid engine coroutine bug
+if _VERSION ~= 'Lua 5.1' or type(jit) == 'table' then
+    local send, comp = wrapRepl()
+    local v = {
+        comp = function(...) return view(comp(...)) end,
+        send = function(...) return view(send(...)) end,
+    }
+    send('(local [foo foo-bar* moe-larry] [1 2 {:*curly* "Why soitenly!"}])')
+    send('(local [!x-y !x_y] [1 2])')
+
+    for _, results in ipairs{
+        {v.comp'foo',  view{'foo', 'foo-bar*'},
+            'local completion works & accounts for mangling'},
+        {v.comp'moe-larry.', '[ "moe-larry.*curly*" ]',
+            'completion traverses tables without mangling keys when input is "tbl-var."'},
+        {v.send'(values !x-y !x_y)', view{{1, 2}},
+            'mangled locals do not collide'},
+        {v.comp'!x', view{'!x-y', '!x_y'},
+            'completions on mangled locals do not collide'},
+    }
+    do
```

Comment:
What's the purpose of the extra `do/end` here?

Reply:
Just to provide scope isolation for the transient values (such as the wrapped REPL instance, which is ended by the `nil` write). Now that everything is wrapped in an `if... end` it's unnecessary, so I'll remove it.

Reply:
Oh wait, no - in this case it's now part of the for loop that runs the test cases. Should I move the `do` up to the same line as the closing brace to make that clearer?

Reply:
I guess it's just that we don't typically worry much about isolating these scopes ; if we re-declare a local with the same name later on, it'll get caught by luacheck, so it just stands out a bit here. But it's fine.